fission-core/library/Fission/Text/Encoded/Class.hs

```diff
@@ -0,0 +1,29 @@
+module Fission.Text.Encoded.Class (ToEncoded (..)) where
+
+import           RIO
+import qualified RIO.ByteString.Lazy as Lazy
+
+import           Fission.Text.Encoded.Types
+import           Fission.Text.Encoding.Types
+
+-- | A helper for beinging some native text into the Encoded type and tagging
+--   it with that text type's correct default encoding
+class ToEncoded txt where
+  type NativeEncoding txt :: Encoding
+  trackEncoding :: txt -> NativeEncoding txt `Encoded` txt
+
+instance ToEncoded Text where
+  type NativeEncoding Text = 'UTF8
+  trackEncoding txt = Encoded txt
+
+instance ToEncoded String where
+  type NativeEncoding String = 'Octal
```

Comment:
I find it weird to think of native encoding here. Is it really octal?

I think of encodings like restrictions on what bytes ranges or byte sequences can appear in a byte string. E.g. binary means bytes can only ever be "0" or "1" (in ascii?), and base64 means only A-Za-z0-9 and some other ascii bytes.

And then by extension wouldn't octal mean that only 0-7 ascii would be accepted bytes in the encoding?

I'm thinking there should be something akin to "all bytes can appear" encoding type.

I might be misunderstanding the interpretation of `Encoded` though, so please let me know :)

Reply:
Ah, you're right — I've seen c2w around, but it does truncate. I'll change that to `Unicode` 👍

Reply:
> I might be misunderstanding the interpretation of Encoded though, so please let me know :)

Yes, possibly. We're saying that "despite the ability to express whatever native characters, you're only using the (e.g.) ASCII character set". We need a way to lift an untagged string type to a tagged version, and show its native encoding. `ByteStrings` are not UTF8 safe, but `Text` is, for instance. Both of them can be used to represent ASCII. Some encodings don't map cleanly onto each other, like Base58-BTC, or Base64 and Base64-URL.

Reply:
> A chinese character:      汉
> it's unicode value:       U+6C49
> convert 6C49 to binary:   01101100 01001001
> embed 6C49 as UTF-8:      11100110 10110001 10001001

😫

Reply:
```haskell
data Text = Text
    {-# UNPACK #-} !A.Array          -- payload (Word16 elements)
    {-# UNPACK #-} !Int              -- offset (units of Word16, not Char)
    {-# UNPACK #-} !Int              -- length (units of Word16, not Char)
    deriving (Typeable)
```

Cool cool cool cool cool

Reply:
https://guide.aelve.com/haskell/strings-o62hqc69

> Text
> [...]
> Uses UTF-16

> String
> [...]
> Doesn't support Unicode perfectly (if you do something like map toUpper, for instance)

![](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fi.imgflip.com%2Fxi8cf.jpg&f=1&nofb=1)

Reply:
Hmm okay, I think I was also letting my Elixir show a bit with `type NativeEncoding ByteString = 'Octal`. That's true of an _unpacked_ `ByteString` (i.e. `[Word8]`) in Haskell, but `ByteString` as a text representation is really ASCII by default. We don't actually use binary, hex, or octal, so I've dropped them in favour of ASCII, Unicode, and so on.

Reply:
Oh wow I totally missed this train of thought when scrolling through here! :D

Anyway, I like the changed version better :)